### PR TITLE
Estimates and routerArray

### PR DESCRIPTION
**Pull request by GainsGoblin**
Fixes:
- Fixed a lot of swap path logic in strategies.
- Added harvest estimations to all strategies.
- BaseStrategy now uses routerArray to swap between routers during harvest.
- SetRouter now can modify routerArray and approves all needed tokens for all routers.
- Harvester now harvests using _estimates array instead of individual parameters.

### CODE REVIEW

**Review by uN2RVw5q — COMMENTED** [review]
Some general comments, the most important one is issues in computing estimates and lack of tests for `getEstimates`.

**Comment by gpersoon** [comment]
At the _harvest functions:
Either use arrayCounter everywhere were applicable (more readable, but slightly more gas)
Or (if you really want to save gas):
Add comments where arrayCounter is replaced with constants (e.g. 0 / i+1)
After the _extraRewardsLength loop, arrayCounter can be calculated directly:
arrayCounter = _extraRewardsLength + 1;
and no need to increase arrayCounter in the loop;
The last increase "arrayCounter += 2;" for example in ETHConvexStrategy.sol isn't necessary

**Comment by gpersoon** [comment]
Why isn't getEstimates() using safemath consistently? (now it's mix of mul / sub / div / - )

**Comment by gpersoon** [comment]
I don't think the following constructions in getEstimates() are valid:
function getEstimates() public view returns (uint256[] memory _estimates) {
    _estimates[0]= ...
    _estimates[_estimates.length] = ...
    
I can't get them working in remix. It does compile but I get an invalid opcode when executing the code.
I think you should allocate memory first and then index the items, something like:
 _estimates = new uint[](crvRewards.extraRewardsLength() + 5);

**Comment by gpersoon** [comment]
in getEstimates()
_slippage is always used as: ONE_HUNDRED_PERCENT - _slippage
might as well calculate that directly to save some gas:
uint256 _notSlippage = ONE_HUNDRED_PERCENT - IHarvester(manager.harvester()).slippage();

File: test/v3/ConvexStrategy3Crv.test.js [hunk 1]
```diff
@@ -53,6 +55,9 @@ describe('ConvexStrategy3Crv', () => {
         unirouter = await ethers.getContractAt('MockUniswapRouter', router.address);
         const vaultPID = 0;
 
+        const harvester = await deployments.get('Harvester');
+        manager.connect(deployer).setHarvester(harvester.address);
```
Comment (transferAndCall) [inline test/v3/ConvexStrategy3Crv.test.js:59]:
```suggestion
        await manager.connect(deployer).setHarvester(harvester.address);
```

File: test/v3/GeneralConvexStrategy.test.js [hunk 2]
```diff
@@ -47,6 +49,9 @@ describe('GeneralConvexStrategy with 2 tokens', () => {
         unirouter = await ethers.getContractAt('MockUniswapRouter', router.address);
         const vaultPID = 0;
 
+        const harvester = await deployments.get('Harvester');
+        manager.connect(deployer).setHarvester(harvester.address);
```
Comment (transferAndCall) [inline test/v3/GeneralConvexStrategy.test.js:53]:
```suggestion
        await manager.connect(deployer).setHarvester(harvester.address);
```

File: test/v3/MIMConvexStrategy.test.js [hunk 3]
```diff
@@ -44,12 +47,20 @@ describe('MIMConvexStrategy', () => {
...
+            MockStableSwap3Pool.address
+        );
         const Controller = await deployments.get('Controller');
         controller = await ethers.getContractAt('Controller', Controller.address);
         const router = await deployments.get('MockUniswapRouter');
         unirouter = await ethers.getContractAt('MockUniswapRouter', router.address);
         const vaultPID = 0;
 
+        const harvester = await deployments.get('Harvester');
+        manager.connect(deployer).setHarvester(harvester.address);
```
Comment (transferAndCall) [inline test/v3/MIMConvexStrategy.test.js:62]:
```suggestion
        await manager.connect(deployer).setHarvester(harvester.address);
```

File: test/v3/NativeStrategyCurve3Crv.test.js [hunk 4]
```diff
@@ -52,6 +54,9 @@ describe('NativeStrategyCurve3Crv', () => {
         const router = await deployments.get('MockUniswapRouter');
         unirouter = await ethers.getContractAt('MockUniswapRouter', router.address);
 
+        const harvester = await deployments.get('Harvester');
+        manager.connect(deployer).setHarvester(harvester.address);
```
Comment (transferAndCall) [inline test/v3/NativeStrategyCurve3Crv.test.js:58]:
```suggestion
        await manager.connect(deployer).setHarvester(harvester.address);
```

File: contracts/v3/strategies/ETHConvexStrategy.sol [hunk 5]
```diff
@@ -82,55 +110,100 @@ contract ETHConvexStrategy is BaseStrategy {
...
 
-            if (balanceOfWant() > 0) {
-                _deposit();
+    function getEstimates() external view returns (uint256[] memory) {
+    	uint256 rewardsLength = crvRewards.extraRewardsLength();
+    	uint256[] memory _estimates = new uint256[](rewardsLength.add(4));
+      	address[] memory _path;
+     	uint256[] memory _amounts;
+      	uint256 _notSlippage = ONE_HUNDRED_PERCENT.sub(IHarvester(manager.harvester()).slippage());
+      	uint256 wethAmount;
```
Comment (transferAndCall) [inline contracts/v3/strategies/ETHConvexStrategy.sol:152]: Please fix this formatting.

**Review by transferAndCall — CHANGES_REQUESTED** [review]

File: contracts/v3/strategies/ETHConvexStrategy.sol (hunk 5 above)
Comment (GainsGoblin) [reply contracts/v3/strategies/ETHConvexStrategy.sol]: Fixed formatting.

**Pull request merged**